### ISSUE TITLE
Flattened-parameter optimizers mutate frozen parameters

### ISSUE THREAD

**Issue by kovacoj** [kovacoj1/optimizers#35]
## Summary
`Annealing`, `Metropolis`, and `Genetic` flatten and overwrite all optimizer parameters, including tensors with `requires_grad=False`.

## Why this matters
Freezing part of a model is standard PyTorch workflow. Even though these optimizers are search-based rather than gradient-based, they still subclass `torch.optim.Optimizer`, so users can reasonably expect frozen parameters to remain unchanged.

Current behavior silently mutates those frozen parameters.

## Affected files
- `src/optimizers/Annealing.py`
- `src/optimizers/Metropolis.py`
- `src/optimizers/Genetic.py`

### Annealing
```python
import torch
from optimizers import Annealing

x = torch.nn.Parameter(torch.tensor([2.0]), requires_grad=False)
y = torch.nn.Parameter(torch.tensor([1.0]))
opt = Annealing([x, y])

before = (x.item(), y.item())
opt.step(lambda: (y ** 2).sum())
after = (x.item(), y.item())
print(before, after)
```
Observed on current `main`:
- before: `(2.0, 1.0)`
- after: `(1.9402999877929688, 1.0543376207351685)`

### Metropolis (deterministic acceptance path)
```python
import torch
from optimizers import Metropolis

x = torch.nn.Parameter(torch.tensor([2.0]), requires_grad=False)
y = torch.nn.Parameter(torch.tensor([1.0]))
opt = Metropolis([x, y])
opt.mutate = lambda: torch.tensor([0.0, 1.0])

orig_rand = torch.rand
torch.rand = lambda *args, **kwargs: torch.tensor([0.0])
try:
    opt.step(lambda: (y ** 2).sum())
finally:
    torch.rand = orig_rand

print(x.item(), y.item())
```
Observed result:
- `x` changes from `2.0` to `0.0`

### Genetic (deterministic population)
```python
import torch
from optimizers import Genetic

x = torch.nn.Parameter(torch.tensor([2.0]), requires_grad=False)
y = torch.nn.Parameter(torch.tensor([1.0]))
opt = Genetic([x, y])
opt.pop_size = 4
opt.population = torch.tensor([[0.0, 1.0]]).repeat(opt.pop_size, 1)

opt.step(lambda: (y ** 2).sum())
print(x.item(), y.item())
```
Observed result:
- `x` changes from `2.0` to `0.0`

## Expected behavior
These optimizers should either:
- ignore frozen parameters consistently, or
- reject them explicitly with a clear error

The current behavior silently mutates values that the user marked as frozen.

**Comment by kovacoj** [comment]
Reopening because PR #37 merged a broken tree:  still mutates frozen parameters in the flattened-parameter optimizers, and CI on the merge commit failed with missing helper wiring. I have a corrected follow-up fix prepared from current main.

**Comment by kovacoj** [comment]
Corrective fix is now in PR #38. This follows the broken merge from PR #37: the helper wiring on  was incomplete, so CI failed and the flattened-parameter optimizers still did not consistently honor frozen parameters. PR #38 restores the intended trainable-only behavior and passes the full test suite locally.
